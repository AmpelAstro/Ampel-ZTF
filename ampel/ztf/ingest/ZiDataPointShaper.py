#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File:                Ampel-ZTF/ampel/ztf/ingest/ZiDataPointShaper.py
# License:             BSD-3-Clause
# Author:              valery brinnel <firstname.lastname@gmail.com>
# Date:                14.12.2017
# Last Modified Date:  10.05.2021
# Last Modified By:    valery brinnel <firstname.lastname@gmail.com>

from typing import Any, Iterable, Sequence
from bson import encode
from ampel.util.hash import hash_payload
from ampel.base.AmpelBaseModel import AmpelBaseModel
from ampel.types import StockId, Tag
from ampel.abstract.AbsT0Unit import AbsT0Unit
from ampel.content.DataPoint import DataPoint
from ampel.ztf.ingest.tags import tags

class ZiDataPointShaperBase(AmpelBaseModel):
	"""
	This class 'shapes' datapoints in a format suitable
	to be saved into the ampel database
	"""

	#: JD2017 is used to define upper limits primary IDs
	JD2017: float = 2457754.5
	#: Byte width of datapoint ids
	digest_size: int = 8

	# Mandatory implementation
	def process(self, arg: Iterable[dict[str, Any]], stock: StockId) -> list[DataPoint]: # type: ignore[override]
		"""
		:param arg: sequence of unshaped pps
		IMPORTANT:
		1) This method *modifies* the input dicts (it removes 'candid' and programpi),
		even if the unshaped pps are ReadOnlyDict instances
		2) 'stock' is not set here on purpose since it will conflict with the $addToSet operation
		"""

		ret_list: list[DataPoint] = []
		setitem = dict.__setitem__
		popitem = dict.pop

		for photo_dict in arg:

			base_tags = tags[photo_dict['programid']][photo_dict['fid']]

			is_prv_candidate = popitem(photo_dict, '_is_prv_candidate', False)

			# Photopoint
			if photo_dict.get('candid'):

				# Cut path if present
				if photo_dict.get('pdiffimfilename'):
					setitem(
						photo_dict, 'pdiffimfilename',
						photo_dict['pdiffimfilename'] \
							.split('/')[-1] \
							.replace('.fz', '')
					)

				# Remove programpi (redundant with programid)
				popitem(photo_dict, 'programpi', None)

				ret_list.append(
					self._create_datapoint(
						stock,
						(["ZTF_ALERT"] if not is_prv_candidate else []) + ["ZTF_DP"],
						photo_dict
					)
				)

			elif "forcediffimflux" in photo_dict:
				
				ret_list.append(self._create_datapoint(stock, ["ZTF_FP"], photo_dict))

			else:

				ret_list.append(
					self._create_datapoint(
						stock,
						["ZTF_UL"],
						{
							'jd': photo_dict['jd'],
							'diffmaglim': photo_dict['diffmaglim'],
							'rcid': (
								rcid
								if (rcid := photo_dict.get('rcid')) is not None
								else (photo_dict['pid'] % 10000) // 100
							),
							'fid': photo_dict['fid'],
							'programid': photo_dict['programid']
						}
					)
				)

		return ret_list
	
	def _create_datapoint(self, stock: StockId, tag: Sequence[Tag], body: dict[str, Any]) -> DataPoint:
		"""
		Create a Datapoint from stock, body, and tags, using the hash of the body as id
		"""
		# ensure that keys are ordered
		sorted_body = dict(sorted(body.items()))
		return {
			"id": hash_payload(encode(sorted_body), size=-self.digest_size*8),
			"stock": stock,
			"tag": [*tags[body['programid']][body['fid']], *tag],
			"body": sorted_body,
		}


class ZiDataPointShaper(ZiDataPointShaperBase, AbsT0Unit):
	
	def process(self, arg: Any, stock: None | StockId = None) -> list[DataPoint]:
		assert stock is not None
		return super().process(arg, stock)
