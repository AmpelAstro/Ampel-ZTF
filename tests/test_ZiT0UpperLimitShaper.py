import pytest
import random

from ampel.ztf.ingest.ZiDataPointShaper import ZiDataPointShaperBase
from ampel.ztf.alert.ZiAlertSupplier import ZiAlertSupplier

# metafixture as suggested in https://github.com/pytest-dev/pytest/issues/349#issuecomment-189370273
@pytest.fixture(params=["upper_limit_alerts", "forced_photometry_alerts"])
def alerts(request):
    yield request.getfixturevalue(request.param)


def test_id_stability(alerts: ZiAlertSupplier, snapshot):
    shaper = ZiDataPointShaperBase()
    alert = next(alerts)
    # NB: process a copy, as ZiDataPointShaper mutates its input
    dps = shaper.process([dict(dp) for dp in alert.datapoints[:5]], alert.stock)
    datapoint_ids = [[dp["tag"][-1], dp["id"]] for dp in dps]
    assert datapoint_ids == snapshot, "datapoint ids are deterministic"

    shuffled_dps = shaper.process(
        [dict(random.sample(list(dp.items()), len(dp))) for dp in alert.datapoints[:5]],
        alert.stock + 1,
    )
    assert [
        [dp["tag"][-1], dp["id"]] for dp in shuffled_dps
    ] == datapoint_ids, "datapoint ids depend only on body content"
