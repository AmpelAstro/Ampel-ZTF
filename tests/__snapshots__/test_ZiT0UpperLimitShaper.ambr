# name: test_id_stability[forced_photometry_alerts]
  <class 'list'> [
    <class 'list'> [
      'ZTF_DP',
      6900491799397119246,
    ],
    <class 'list'> [
      'ZTF_FP',
      -1755349362406496774,
    ],
    <class 'list'> [
      'ZTF_FP',
      -3563520953961907462,
    ],
    <class 'list'> [
      'ZTF_FP',
      -2251392325347245153,
    ],
    <class 'list'> [
      'ZTF_FP',
      -2060288752642886212,
    ],
  ]
---
# name: test_id_stability[upper_limit_alerts]
  <class 'list'> [
    <class 'list'> [
      'ZTF_DP',
      -5869826593587407452,
    ],
    <class 'list'> [
      'ZTF_UL',
      -1244258051360039709,
    ],
    <class 'list'> [
      'ZTF_UL',
      -519848071864680703,
    ],
    <class 'list'> [
      'ZTF_UL',
      3085705785494620878,
    ],
    <class 'list'> [
      'ZTF_UL',
      5968690154336543417,
    ],
  ]
---
